[DOCS] DOC-1180 - Add documentation for Custom Actions in GX Cloud (#11521)
great-expectations/great_expectations: docs/docusaurus/docs/cloud/alerts/custom_actions.md

> ---

Reviewer: This new page should be added to the [Respond to results](https://deploy-preview-11521.docs.greatexpectations.io/docs/cloud/alerts/alerts_lp) landing page

sidebar_label: 'Custom actions'

> title: 'Custom actions'

Reviewer: Because this is instructional content, as opposed to conceptual, it should ideally have an imperative in the title.

```suggestion
sidebar_label: 'Create a custom action'
title: 'Create a custom action'
```

description: Run custom logic based on Validation Results.
---

import TabItem from '@theme/TabItem';
import Tabs from '@theme/Tabs';

> Great Expectations provides [Actions for common workflows](cloud/alerts/trigger_actions.md) such as sending emails and sending Microsoft Teams notifications. If these don't meet your needs, you can create a custom Action to integrate with different tools, such as Atlassian Jira, or apply custom business logic based on Validation Results. Example use cases for custom Actions include:

Reviewer: suggestion for grammar and concision

```suggestion
Great Expectations provides [Actions for common workflows](cloud/alerts/trigger_actions.md), such as sending emails and Microsoft Teams notifications. If these don't meet your needs, you can create a custom Action to integrate with different tools, such as Atlassian Jira, or apply custom business logic based on Validation Results. Example use cases for custom Actions include:
```

   ```

3.  Optional. Add any additional fields your Action requires at runtime. Actions are built on Pydantic models. Define the field name as a class-level attribute on your Action, and annotate it with the correct type. When you instantiate the Action, pass the field value into the Action `init` method. Your Action will have access to these values within your custom `run` method through `self.<MY_FIELD_NAME>`. The example below shows how to include a field for accessing a Jira API.

      ```python title="Python" name="docs/docusaurus/docs/cloud/alerts/_examples/create_a_custom_action.py - add custom fields"
      ```

4. Override the `run()` method with the logic for the Action. The example below shows how to create a Jira ticket using the additional field you added in the previous step.

   ```python title="Python" name="docs/docusaurus/docs/cloud/alerts/_examples/create_a_custom_action.py - override run" 
   ```

</TabItem>

<TabItem value="sample_code" label="Sample code">

```python title="Python" name="docs/docusaurus/docs/cloud/alerts/_examples/create_a_custom_action.py - full code example" 
```

</TabItem>

</Tabs>

> Now you can use your custom Action like you would any built-in Action. [Trigger your custom Action](cloud/alerts/trigger_actions.md) to start automating responses to Validation Results.

Reviewer: As discussed on slack ([DM](https://greatexpectationslabs.slack.com/archives/D07NADFQTME/p1763504785887849) and [public channel](https://greatexpectationslabs.slack.com/archives/C013GAN00GM/p1763565321267109)) it seems like there's either a limitation regarding UI-ran validations or some extra steps required for UI-ran validations to successfully trigger a custom action. This paragraph should be clarified one way or the other once we get confirmation about whether or not there's a workaround.

Note that I only tested ad hoc validations with the validate button. I did not yet test scheduled validations. I assume that scheduled validations have the same behavior as the validate button, but it may be prudent to test that to confirm once we know whether or not there's a workaround for UI validations.

---
sidebar_label: 'Create a custom action'
title: 'Create a custom action'
description: Run custom logic based on Validation Results.
---

import TabItem from '@theme/TabItem';
import Tabs from '@theme/Tabs';

> Great Expectations provides [Actions for common workflows](cloud/alerts/trigger_actions.md) such as sending emails and Microsoft Teams notifications. If these don't meet your needs, you can create a custom Action to integrate with different tools, such as Atlassian Jira, or apply custom business logic based on Validation Results. Example use cases for custom Actions include:

Reviewer: I still think there should be a comma before the first "such as". Is there a reason you skipped that part of my [original committable suggestion for this paragraph](https://github.com/great-expectations/great_expectations/pull/11521#discussion_r2542600366) while you implemented the other part of that suggestion?

```suggestion
Great Expectations provides [Actions for common workflows](cloud/alerts/trigger_actions.md), such as sending emails and Microsoft Teams notifications. If these don't meet your needs, you can create a custom Action to integrate with different tools, such as Atlassian Jira, or apply custom business logic based on Validation Results. Example use cases for custom Actions include:
```

Author: I interpreted this as a stylistic suggestion, but I'm fine either way.

Reviewer: In the future, if you intentionally skip addressing some feedback or implementing part of a suggestion because you disagree with it or don't think it's worth the effort or any other reason I would appreciate if you would please note that you're skipping it intentionally and explain the reason. Otherwise I will think you skipped it on accident and repeat the feedback in my re-review.

Thanks for adding the comma. I saw this as objective grammar feedback, not subjective style feedback. In this case, "such as" is introducing a nonrestrictive clause so there should be a comma before it

Author: Understood and will do for the future.

</Tabs>

> Now you can use your custom Action like you would any built-in Action through the GX Cloud API. [Trigger your custom Action](cloud/alerts/trigger_actions.md) to start automating responses to Validation Results.

Reviewer: The first sentence here is grammatically correct but the flow feels a bit backwards to me. 

It's structured as [user action] [comparison action] [interface for user action]
I think it would flow more smoothly as  [user action] [interface for user action] [comparison action]

```suggestion
Now you can use your custom Action through the GX Cloud API like you would any built-in Action. [Trigger your custom Action](cloud/alerts/trigger_actions.md) to start automating responses to Validation Results.
```

## Limitations

Keep the following limitations in mind when using custom Actions.

> - When using a [fully hosted deployment](cloud/deploy/deployment_patterns.md#fully-hosted-deployment), custom Actions cannot be executed through the UI

Reviewer: typo

```suggestion
- When using a [fully-hosted deployment](cloud/deploy/deployment_patterns.md#fully-hosted-deployment), custom Actions cannot be executed through the UI
```

> - When using a [agent-enabled deployment](cloud/deploy/deployment_patterns.md#agent-enabled-deployment), custom Actions can be executed through the UI by [adding custom commands to GX Agent image](cloud/deploy/deploy_gx_agent.md#deploy-the-gx-agent)

Reviewer: Suggestion below addresses the following feedback, some of which is objective (typos) and some of which is subjective (phrasing). Feel free to address the subjective feedback in a different way but please make sure to consider all aspects of the feedback

1. "fully hosted" should have a hyphen because it's a compound adjective in this context
2. "a agent-enabled" should be "an agent-enabled" for proper phonetic article choice
3. I think these bullet points should end with periods similar to the first unordered list on this page 
4. "adding custom commands to GX Agent image" is missing an article (... to the GX Agent image...)
5. I find "executed through the UI" to be kind of odd wording here. To me this phrasing implies that there's a button in the UI that directly runs the action. But, it's more that clicking the validate button in the UI, or configuring a schedule in the UI, runs one or more validations that can subsequently trigger built-in actions but not custom actions.
6. Taking a closer look at the behavior I'm seeing when trying to run a validation in the UI, it's not just that the custom action won't be triggered, it's that the validation won't run at all. 
7. The phrasing of the second bullet point doesn't feel like it fits the framing of "limitation" - it's written in terms of what you can do, not what you can't do. 

```suggestion
- When using a [fully-hosted deployment](cloud/deploy/deployment_patterns.md#fully-hosted-deployment), if you add a custom Action to your GX-managed checkpoint, you will no longer be able to use the **Validation schedule** or the **Validate** button in the UI. 
- When using an [agent-enabled deployment](cloud/deploy/deployment_patterns.md#agent-enabled-deployment),  if you add a custom Action to your GX-managed checkpoint, you will no longer be able to use the **Validation schedule** or the **Validate** button in the UI unless you [customize the GX Agent image](cloud/deploy/deploy_gx_agent.md#deploy-the-gx-agent) with a command for your custom Action.
```
